File: scripts/diagnostics/search3_direct_anex_runtime_probe.php
<?php
declare(strict_types=1);

/** Installed SEARCH boundary diagnosis only; never installs code or calls a quote. */
const S3_ANEX_OPERATION = 'search3-direct-anex-runtime-probe-20260926-v1';
const S3_ANEX_SOURCE = '93ce2190a444a7b4aa2920a013f14c8b2f64e79c';
const S3_ANEX_BLOBS = [
    'api-anex-search3-preview.php' => '22710804d8ae108098e3c545efe9835ff14e0024',
    'app/integrations/anex-client.php' => 'b1e7691409bf7289ca74e9562829dddde5c76188',
    'app/integrations/anex-search.php' => '51546f5fcc5d6f471a391e580a87d6137ea2256a',
];
function s3_anex_error(Throwable $error, array $last): array {
    $known = ['ANEX_TOKEN_REQUIRED','ANEX_REQUEST_LIMIT','ANEX_TRANSPORT_ERROR','ANEX_INVALID_RESPONSE',
        'ANEX_RESPONSE_TOO_LARGE','ANEX_HTTP_ERROR','ANEX_SUPPLIER_ERROR','ANEX_INVALID_PARAMS',
        'ANEX_INVALID_PRICES','ANEX_SEARCH_PAGINATION_LIMIT','ANEX_RATE_LIMIT',
        'ANEX_FILTER_UNSUPPORTED','ANEX_DESTINATION_UNSUPPORTED','ANEX_INVALID_SEARCH',
        'project_invalid','runtime_missing','runtime_drift','config_unavailable','operation_exists_no_replay',
        'reservation_failed','scope_expired','request_budget_exceeded'];
    $class = get_class($error);
    $safe = ['exception' => in_array($class, ['PDOException','RuntimeException','InvalidArgumentException','Error','TypeError'], true) ? $class : 'other',
        'reason' => in_array($error->getMessage(), $known, true) ? $error->getMessage() : 'internal_error'];
    $actions = ['SearchTour_TOWNFROMS','SearchTour_STATES','SearchTour_CURRENCIES','SearchTour_PRICES'];
    if (in_array($last['action'] ?? null, $actions, true)) $safe['action'] = $last['action'];
    foreach (['http_status'=>599,'curl_errno'=>999,'supplier_code'=>99999,'response_bytes'=>2097153] as $key=>$max) {
        if (isset($last[$key]) && is_int($last[$key]) && $last[$key] >= 0 && $last[$key] <= $max) $safe[$key] = $last[$key];
    }
    if ($error instanceof PDOException && preg_match('/\A[A-Z0-9]{5}\z/D', (string)$error->getCode())) $safe['sql_state'] = (string)$error->getCode();
    return $safe;
}
function s3_anex_durable(string $path, array $value): void {
    $body = json_encode($value, JSON_THROW_ON_ERROR);
    $f = @fopen($path, 'x');
    if ($f === false) throw new RuntimeException('operation_exists_no_replay');
    try {
        if (!chmod($path, 0600) || fwrite($f, $body) !== strlen($body) || !fflush($f)
            || (function_exists('fsync') && !fsync($f))) throw new RuntimeException('reservation_failed');
    } finally { fclose($f); }
}
function s3_anex_verify(string $target): void {
    foreach (S3_ANEX_BLOBS as $name=>$expected) {
        $path=$target.'/'.$name;
        if (!is_file($path) || is_link($path) || filesize($path)>2097152) throw new RuntimeException('runtime_missing');
        $bytes=file_get_contents($path);
        if (!is_string($bytes) || sha1('blob '.strlen($bytes)."\0".$bytes)!==$expected) throw new RuntimeException('runtime_drift');
    }
}
function search3_direct_anex_runtime_main(): int {
    error_reporting(0); ini_set('display_errors','0'); ini_set('log_errors','0');
    $result=['operation'=>S3_ANEX_OPERATION,'source'=>S3_ANEX_SOURCE,'status'=>'blocked','stage'=>'preflight',
        'client_requests'=>0,'database_writes'=>0,'quote_calls'=>0,'replay_allowed'=>false];
    $pdo=null; $client=null; $dir=null; $stage='preflight';
    try {
        $root=realpath(getcwd());
        if (PHP_SAPI!=='cli' || !$root || basename($root)!=='anytoour.ru') throw new RuntimeException('project_invalid');
        $target=$root.'/_preview/search3-anex-candidate';
        $private=dirname($root,2).'/.anytoour-anex';
        if (!is_dir($target) || is_link($target) || !is_dir($private) || is_link($private)) throw new RuntimeException('runtime_missing');
        s3_anex_verify($target);
        if ((new DateTimeImmutable('today', new DateTimeZone('Europe/Moscow')))->format('Y-m-d')>'2026-10-10') throw new RuntimeException('scope_expired');
        $operationDir=$private.'/'.S3_ANEX_OPERATION;
        if (file_exists($operationDir) || is_link($operationDir)) throw new RuntimeException('operation_exists_no_replay');
        if (!mkdir($operationDir,0700)) throw new RuntimeException('reservation_failed');
        $dir=$operationDir;
        s3_anex_durable($dir.'/reservation.json',['operation'=>S3_ANEX_OPERATION,'source'=>S3_ANEX_SOURCE,'at'=>time(),'replay_allowed'=>false]);
        $stage='runtime_include';
        $_SERVER['DOCUMENT_ROOT']=$root;
        require_once $target.'/api-anex-search3-preview.php';
        require_once $target.'/app/integrations/anex-initial-week-gate.php';
        require_once $target.'/app/integrations/anex-preview-gateway.php';
        require_once $target.'/app/integrations/anex-search-mapping-registry.php';
        if (!is_file($target.'/.anex-private.php')) throw new RuntimeException('config_unavailable');
        require_once $target.'/.anex-private.php';
        $token=trim((string)getenv('ANEX_API_TOKEN'));
        if ($token==='' && defined('ANEX_API_TOKEN')) $token=trim((string)ANEX_API_TOKEN);
        if ($token==='') throw new RuntimeException('config_unavailable');
        $stage='database';
        $helper=is_file($root.'/data/db-v1.php')?$root.'/data/db-v1.php':$root.'/v2/data/db-v1.php';
        require_once $helper;
        $pdo=v2_data_db();
        $pdo->exec('START TRANSACTION READ ONLY');
        $client=new AnyTourAnexClient($token); unset($token);
        $params=['departureId'=>'1','countryId'=>'4','dateFrom'=>'2026-10-10','dateTo'=>'2026-10-16',
            'nightsFrom'=>7,'nightsTo'=>7,'adults'=>2,'childs'=>[],'meal'=>'','hotelCategory'=>'','hotelRating'=>'',
            'hotelTypes'=>[],'hotelIds'=>[],'hotelServices'=>[],'arrivalId'=>'','regionIds'=>[],'subregionIds'=>[],
            'operatorIds'=>[],'priceFrom'=>'','priceTo'=>'','currency'=>'RUB','onlyCharter'=>false,'onlyDirect'=>false];
        $cache=[]; $diagnostics=[]; $state=[]; $stage='search_core';
        // Exactly one invocation, using the unchanged installed client (hard limit 12).
        // No HTTP handler retry, no observer, no program-observation or offer autosave.
        $data=anytour_anex_search3_run(['action'=>'search','generation'=>1,'params'=>$params],$pdo,$client,$cache,$diagnostics,null,$state,'all');
        $result['hotels']=count($data['hotels']??[]);
        $result['offers']=array_sum(array_map(static fn(array $hotel):int=>count($hotel['tours']??[]),$data['hotels']??[]));
        s3_anex_verify($target);
        $result['status']='complete';
    } catch (Throwable $error) {
        $result['status']=$stage==='search_core'?'failed':'blocked';
        $result['failure']=s3_anex_error($error,$client!==null?$client->lastRequestDiagnostics():[]);
    } finally {
        $result['stage']=$stage;
        if ($client!==null) $result['client_requests']=$client->requestsMade();
        if ($pdo instanceof PDO && $pdo->inTransaction()) $pdo->rollBack();
        if ($dir!==null) {
            try { s3_anex_durable($dir.'/result.json',$result); }
            catch (Throwable $ignored) { $result['status']='unknown'; }
        }
    }
    echo json_encode($result,JSON_THROW_ON_ERROR),"\n";
    return $result['status']==='unknown'?1:0;
}

File: scripts/diagnostics/search3_direct_anex_runtime_probe.py
"""Owner-gated one-shot diagnosis of the fixed installed ANEX search core."""
import json
import os
from pathlib import Path
import re
import subprocess
import sys
import tempfile
import urllib.request

COMMAND = '/search3-direct-anex-runtime-probe-v1'
REPO = 'pyatkoff/poisk-turov-test'
OPERATION = 'search3-direct-anex-runtime-probe-20260926-v1'
SOURCE = '93ce2190a444a7b4aa2920a013f14c8b2f64e79c'

def authorize(event, env):
    issue, comment = event.get('issue', {}), event.get('comment', {})
    return (env.get('GITHUB_EVENT_NAME') == 'issue_comment' and event.get('action') == 'created'
        and env.get('GITHUB_REPOSITORY') == REPO and env.get('GITHUB_REF') == 'refs/heads/main'
        and env.get('GITHUB_ACTOR') == env.get('GITHUB_TRIGGERING_ACTOR') == 'pyatkoff'
        and env.get('GITHUB_RUN_ATTEMPT') == '1' and issue.get('number') == 3419
        and 'pull_request' not in issue and comment.get('body') == COMMAND
        and comment.get('author_association') == 'OWNER'
        and comment.get('user', {}).get('id') == 226193297
        and comment.get('user', {}).get('login') == 'pyatkoff')

def sanitize(value):
    if not isinstance(value, dict) or value.get('operation') != OPERATION or value.get('source') != SOURCE:
        raise ValueError('invalid_receipt_identity')
    clean = {'operation': OPERATION, 'source': SOURCE}
    for key, allowed in {'status': {'blocked','failed','complete','unknown'},
                         'stage': {'preflight','runtime_include','database','search_core'}}.items():
        if value.get(key) not in allowed: raise ValueError('invalid_receipt_enum')
        clean[key] = value[key]
    for key, cap in {'client_requests':12,'database_writes':0,'quote_calls':0,'hotels':4800,'offers':4800}.items():
        if key not in value and key in {'hotels','offers'}: continue
        if type(value.get(key)) is not int or not 0 <= value[key] <= cap: raise ValueError('invalid_receipt_count')
        clean[key] = value[key]
    if value.get('replay_allowed') is not False: raise ValueError('invalid_receipt_replay')
    clean['replay_allowed'] = False
    failure = value.get('failure')
    if isinstance(failure, dict):
        out = {}
        for key, allowed in {
            'exception': {'PDOException','RuntimeException','InvalidArgumentException','Error','TypeError','other'},
            'action': {'SearchTour_TOWNFROMS','SearchTour_STATES','SearchTour_CURRENCIES','SearchTour_PRICES'},
            'reason': {'ANEX_TOKEN_REQUIRED','ANEX_REQUEST_LIMIT','ANEX_TRANSPORT_ERROR','ANEX_INVALID_RESPONSE',
                'ANEX_RESPONSE_TOO_LARGE','ANEX_HTTP_ERROR','ANEX_SUPPLIER_ERROR','ANEX_INVALID_PARAMS',
                'ANEX_INVALID_PRICES','ANEX_SEARCH_PAGINATION_LIMIT','ANEX_RATE_LIMIT','ANEX_FILTER_UNSUPPORTED',
                'ANEX_DESTINATION_UNSUPPORTED','ANEX_INVALID_SEARCH','project_invalid','runtime_missing',
                'runtime_drift','config_unavailable','operation_exists_no_replay','reservation_failed',
                'scope_expired','request_budget_exceeded','internal_error'},
        }.items():
            if isinstance(failure.get(key), str) and failure[key] in allowed: out[key] = failure[key]
        for key, cap in {'http_status':599,'curl_errno':999,'supplier_code':99999,'response_bytes':2097153}.items():
            if type(failure.get(key)) is int and 0 <= failure[key] <= cap: out[key] = failure[key]
        state = failure.get('sql_state')
        if isinstance(state, str) and re.fullmatch('[A-Z0-9]{5}', state): out['sql_state'] = state
        clean['failure'] = out
    return clean

def main():
    event=json.loads(Path(os.environ['GITHUB_EVENT_PATH']).read_text())
    if not authorize(event, os.environ): raise ValueError('authorization_denied')
    if sys.argv[1:] == ['--authorize']:
        req=urllib.request.Request(f"https://api.github.com/repos/{REPO}/issues/comments/{event['comment']['id']}",
            headers={'Authorization':'Bearer '+os.environ['GH_TOKEN'],'Accept':'application/vnd.github+json'})
        with urllib.request.urlopen(req,timeout=20) as response: fresh=json.load(response)
        if not authorize({**event,'comment':fresh},os.environ): raise ValueError('command_revoked')
        print('SEARCH3_ANEX_RUNTIME_AUTHORIZED')
        return
    if sys.argv[1:] != ['--run']: raise ValueError('invalid_mode')
    host, user = os.environ['INT_SSH_HOST'], os.environ['INT_SSH_USER']
    if not re.fullmatch('[A-Za-z0-9][A-Za-z0-9.-]*',host) or not re.fullmatch('[A-Za-z0-9_][A-Za-z0-9_.-]{0,63}',user):
        raise ValueError('invalid_ssh_identity')
    code=Path(__file__).with_suffix('.php').read_text()+'\nexit(search3_direct_anex_runtime_main());\n'
    with tempfile.TemporaryDirectory(dir=os.environ['RUNNER_TEMP']) as work:
        key=Path(work)/'key'; key.write_text(os.environ['INT_SSH_KEY'].rstrip()+'\n');key.chmod(0o600)
        args=['ssh','-T','-i',str(key),'-o','IdentitiesOnly=yes','-o','BatchMode=yes',
            '-o','StrictHostKeyChecking=accept-new','-o','UserKnownHostsFile='+str(Path(work)/'known_hosts'),
            '-o','ConnectTimeout=15','-o','ServerAliveInterval=15','-o','ServerAliveCountMax=2','-o','LogLevel=ERROR',
            '-l',user,host,'cd "$HOME/www/anytoour.ru" && php']
        child_env={k:v for k,v in os.environ.items() if k not in {'INT_SSH_KEY','GH_TOKEN'}}
        completed=subprocess.run(args,input=code,text=True,capture_output=True,timeout=300,env=child_env)
        if len(completed.stdout) > 8192: raise ValueError('invalid_remote_output')
        result=sanitize(json.loads(completed.stdout))
        out=Path('search3-direct-anex-runtime-probe');out.mkdir(exist_ok=True)
        (out/'result.json').write_text(json.dumps(result,indent=2)+'\n')
        print(json.dumps(result))
        if completed.returncode or result['status']=='unknown': raise ValueError('operation_unknown_no_replay')

if __name__ == '__main__':
    try: main()
    except Exception:
        # Never print subprocess, HTTP, PHP or token-bearing exception text.
        sys.exit('SEARCH3_ANEX_RUNTIME_UNCONFIRMED_NO_REPLAY')
